File: src/Rezoom.SQL.Mapping/ConfigurationConnectionProvider.fs
namespace Rezoom.SQL.Mapping
open System
open Microsoft.Extensions.Configuration

/// <summary>
/// <see cref="ConnectionProvider"/> that resolves settings via Microsoft.Extensions.Configuration
/// (appsettings.json, environment variables, anything else wired into the host's
/// <see cref="IConfiguration"/>).
/// </summary>
/// <remarks>
/// Looks up:
/// <list type="bullet">
///   <item>Connection string at <c>ConnectionStrings:{name}</c></item>
///   <item>Provider invariant at <c>RezoomSQL:Providers:{name}</c>, defaulting to
///         <c>Microsoft.Data.SqlClient</c> when unset</item>
/// </list>
/// Typical ASP.NET Core registration:
/// <code>
///   services.AddSingleton&lt;ConnectionProvider, ConfigurationConnectionProvider&gt;();
/// </code>
/// </remarks>
type ConfigurationConnectionProvider(configuration : IConfiguration) =
    inherit ConnectionProvider()

    [<Literal>]
    static let DefaultProviderName = "Microsoft.Data.SqlClient"

    override __.GetConnectionString(name) =
        let connectionString = configuration.[sprintf "ConnectionStrings:%s" name]
        if String.IsNullOrEmpty(connectionString) then
            failwithf
                "No connection string named '%s' in configuration (looked for ConnectionStrings:%s)"
                name name
        let providerName =
            match configuration.[sprintf "RezoomSQL:Providers:%s" name] with
            | null | "" -> DefaultProviderName
            | v -> v
        {   Name = name
            ConnectionString = connectionString
            ProviderName = providerName
        }

    override this.Open(name) =
        let info = this.GetConnectionString(name)
        let factory = NetStandardHacks.DbProviderFactories.GetFactory(info.ProviderName)
        let conn = factory.CreateConnection()
        if isNull conn then
            failwithf "Provider '%s' returned a null DbConnection" info.ProviderName
        conn.ConnectionString <- info.ConnectionString
        conn.Open()
        if conn.GetType().Name = "SQLiteConnection" then
            // Encourage SQLite to put the R in RDBMS
            use cmd = conn.CreateCommand()
            cmd.CommandText <- "PRAGMA foreign_keys=ON;"
            cmd.ExecuteNonQuery() |> ignore
        conn

File: demos/SQLFiddle/SQLFiddle.Website/ConfigurationConnectionProvider.cs
using System.Configuration;
using System.Data.Common;
using Rezoom.SQL.Mapping;

namespace SQLFiddle.Website;

/// <summary>
/// Pulls connection settings from ASP.NET Core's <see cref="IConfiguration"/>:
/// the connection string from the standard <c>ConnectionStrings:{name}</c> slot,
/// and the provider invariant (e.g. <c>"Microsoft.Data.SqlClient"</c>) from
/// <c>RezoomSQL:Providers:{name}</c>, falling back to <c>Microsoft.Data.SqlClient</c>.
///
/// Wire it up in Program.cs as a singleton ConnectionProvider; controllers (and
/// the migration call) then take it via constructor injection like any other
/// service. No App.config required.
/// </summary>
public sealed class ConfigurationConnectionProvider(IConfiguration cfg) : ConnectionProvider
{
    public override ConnectionStringSettings GetConnectionString(string name)
    {
        var connectionString = cfg.GetConnectionString(name)
            ?? throw new InvalidOperationException(
                $"No connection string named '{name}' under ConnectionStrings:{name}");
        var providerName = cfg[$"RezoomSQL:Providers:{name}"] ?? "Microsoft.Data.SqlClient";
        return new ConnectionStringSettings(name, connectionString, providerName);
    }

    public override DbConnection Open(string name)
    {
        var settings = GetConnectionString(name);
        var factory = NetStandardHacks.DbProviderFactories.GetFactory(settings.ProviderName);
        var conn = factory.CreateConnection()
            ?? throw new InvalidOperationException(
                $"Provider '{settings.ProviderName}' returned a null DbConnection");
        conn.ConnectionString = settings.ConnectionString;
        conn.Open();
        return conn;
    }
}
